### PR TITLE
Refactor AST for zero-cost modules

### PR DESCRIPTION
**Pull request by Schaeff**
We currently handle imports by compiling (all the way to flat_absy) dependencies and injecting them as "bytecode".

There are limitations as optimisations that are call-specific (ie depend on values of the arguments) cannot be executed on imported function.

Change this to build an AST for the whole dependency tree, do call-agnostic transformations, then inline all calls, then do call-specific optimisations on the inlined result.

This PR refactors quite a few internal data structures to make things simpler and compile-time.

Todo:
- [x] improve performance of inliner (currently does too many String allocations)
- [x] use canonical keys for modules, so that `./../foo/foo.code` is equivalent to `./foo.code` (currently it is not as we're using strings) update: this can be done in a different PR

# Features
- What we know as a program is now a module, and a program is a collection of modules
- Imports from other modules are done by name, so all matching functions are imported
- API changes to the resolvers to work more on &str
- All calls are inlined in the semantic phase, except low level embeds
- sha256_round and split are now under the same `EMBED/` import prefix

### CODE REVIEW

**Comment by JacobEberhardt** [comment]
I think the separation between call-agnostic and specific optimizations is fine and makes sense.
Is this to be future-proof or, are there actually any call-agnostic optimizations that we would now apply before inlining?

**Comment by stefandeml** [comment]
Very well done! Mostly doc and typos. 
Well documented and good to follow. Can't really comment on logical bugs during resolving and synthesizing.

File: zokrates_core/tests/bench/split.code [hunk 1]
```diff
@@ -1,7 +1,7 @@
-import "PACKING/split"
+import "EMBED/unpack"
```
[the remark this answers is not in the document]
Comment (Schaeff) [reply zokrates_core/tests/bench/split.code]: I changed the packing section of the docs to match stdlib, however it does not mention embedded gadgets, and don't think it should? It's implementation detail of the stdlib

File: zokrates_core/src/absy/from_ast.rs [hunk 2]
```diff
@@ -70,6 +69,12 @@ impl<'ast, T: Field> From<pest::Function<'ast>> for absy::FunctionNode<'ast, T>
                 .collect(),
             signature,
         }
+        .span(span.clone()); // TODO check
```
[the remark this answers is not in the document]
Comment (Schaeff) [reply zokrates_core/src/absy/from_ast.rs]: 👍

File: zokrates_core/src/absy/from_ast.rs [hunk 3]
```diff
@@ -552,55 +557,64 @@ mod tests {
         let source = "def main() -> (field): return 42
         ";
```
[the remark this answers is not in the document]
Comment (Schaeff) [reply zokrates_core/src/absy/from_ast.rs]: you're right we don't

File: zokrates_core/tests/bench/split.code (hunk 1 above)
[the remark this answers is not in the document]
Comment (stefandeml) [reply zokrates_core/tests/bench/split.code]: ok, makes sense!

**Pull request merged**